File: API/2653_Martin_Martingale/CS/MartinMartingaleStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Martingale grid that alternates long and short entries while doubling volume.
/// </summary>
public class MartinMartingaleStrategy : Strategy
{
	private readonly StrategyParam<int> _stepPoints;
	private readonly StrategyParam<int> _entryOffsetPoints;
	private readonly StrategyParam<decimal> _profitTarget;
	private readonly StrategyParam<DataType> _candleType;

	private readonly HashSet<long> _processedOrders = new();

	private Order _buyStopOrder;
	private Order _sellStopOrder;
	private Order _reversalOrder;

	private decimal _stepSize;
	private decimal _entryOffset;
	private decimal _lastTradePrice;
	private decimal _lastTradeVolume;
	private int _martingaleLevel;
	private Sides? _lastTradeSide;
	private bool _isClosing;

	/// <summary>
	/// Distance in points that defines when the next reversal is triggered.
	/// </summary>
	public int StepPoints
	{
		get => _stepPoints.Value;
		set => _stepPoints.Value = value;
	}

	/// <summary>
	/// Offset in points for the initial stop orders around the current price.
	/// </summary>
	public int EntryOffsetPoints
	{
		get => _entryOffsetPoints.Value;
		set => _entryOffsetPoints.Value = value;
	}

	/// <summary>
	/// Aggregated profit required to close the entire martingale cycle.
	/// </summary>
	public decimal ProfitTarget
	{
		get => _profitTarget.Value;
		set => _profitTarget.Value = value;
	}

	/// <summary>
	/// Candle type used to monitor the price.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initializes a new instance of <see cref="MartinMartingaleStrategy"/>.
	/// </summary>
	public MartinMartingaleStrategy()
	{
		_stepPoints = Param(nameof(StepPoints), 10)
			.SetGreaterThanZero()
			.SetDisplay("Step (points)", "Distance multiplier for reversals", "General")
			.SetCanOptimize(true);

		_entryOffsetPoints = Param(nameof(EntryOffsetPoints), 10)
			.SetGreaterThanZero()
			.SetDisplay("Entry Offset (points)", "Offset for initial stop orders", "General")
			.SetCanOptimize(true);

		_profitTarget = Param(nameof(ProfitTarget), 5m)
			.SetGreaterThanZero()
			.SetDisplay("Profit Target", "Total profit to close all positions", "Risk")
			.SetCanOptimize(true);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Candles for price monitoring", "Data");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		CancelActiveOrders();
		ResetCycle();
		_isClosing = false;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		UpdateStepSettings();

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	/// <inheritdoc />
	protected override void OnOrderChanged(Order order)
	{
		base.OnOrderChanged(order);

		if (order.State != OrderStates.Done &&
			order.State != OrderStates.Failed &&
			order.State != OrderStates.Cancelled)
		{
			return;
		}

		if (order == _buyStopOrder)
		{
			_buyStopOrder = null;
		}
		else if (order == _sellStopOrder)
		{
			_sellStopOrder = null;
		}
		else if (order == _reversalOrder)
		{
			_reversalOrder = null;
		}
	}

	/// <inheritdoc />
	protected override void OnNewMyTrade(MyTrade trade)
	{
		base.OnNewMyTrade(trade);

		if (trade.Trade is null)
			return;

		var order = trade.Order;
		if (order is null)
			return;

		if (order.TransactionId == 0)
			return;

		if (!_processedOrders.Add(order.TransactionId))
			return;

		if (order != _buyStopOrder && order != _sellStopOrder && order != _reversalOrder)
			return;

		_lastTradePrice = trade.Trade.Price;

		var volume = order.Volume;
		if (volume <= 0m)
		{
			volume = trade.Trade.Volume;
		}

		if (volume > 0m)
		{
			_lastTradeVolume = volume;
		}

		_lastTradeSide = order.Direction;

		if (_martingaleLevel == 0)
		{
			_martingaleLevel = 1;
		}
		else
		{
			_martingaleLevel++;
		}

		if (order == _buyStopOrder && _sellStopOrder != null)
		{
			CancelOrderIfActive(ref _sellStopOrder);
		}
		else if (order == _sellStopOrder && _buyStopOrder != null)
		{
			CancelOrderIfActive(ref _buyStopOrder);
		}
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		UpdateStepSettings();

		if (_stepSize <= 0m || Volume <= 0m)
			return;

		var price = candle.ClosePrice;

		if (_isClosing)
		{
			if (Position == 0)
			{
				_isClosing = false;
				ResetCycle();
			}
			else
			{
				return;
			}
		}

		if (!_isClosing && Position == 0 && _martingaleLevel > 0)
		{
			ResetCycle();
		}

		var totalProfit = CalculateTotalProfit(price);
		if (ProfitTarget > 0m && totalProfit >= ProfitTarget && (Position != 0 || _martingaleLevel > 0))
		{
			_isClosing = true;
			CloseAll();
			CancelActiveOrders();
			return;
		}

		if (_martingaleLevel == 0 && Position == 0)
		{
			EnsureInitialOrders(price);
			return;
		}

		if (_lastTradeSide is null || _martingaleLevel == 0)
			return;

		if (IsOrderActive(_reversalOrder))
			return;

		var threshold = _martingaleLevel * _stepSize;

		if (_lastTradeSide == Sides.Buy)
		{
			if (price <= _lastTradePrice - threshold)
			{
				PlaceReversal(Sides.Sell);
			}
		}
		else
		{
			if (price >= _lastTradePrice + threshold)
			{
				PlaceReversal(Sides.Buy);
			}
		}
	}

	private void EnsureInitialOrders(decimal price)
	{
		if (_entryOffset <= 0m || price <= 0m)
			return;

		if (!IsOrderActive(_buyStopOrder))
		{
			_buyStopOrder = BuyStop(Volume, price + _entryOffset);
		}

		if (!IsOrderActive(_sellStopOrder))
		{
			_sellStopOrder = SellStop(Volume, price - _entryOffset);
		}
	}

	private void PlaceReversal(Sides side)
	{
		var nextVolume = GetNextVolume();
		if (nextVolume <= 0m)
			return;

		_reversalOrder = side == Sides.Buy
			? BuyMarket(nextVolume)
			: SellMarket(nextVolume);
	}

	private decimal GetNextVolume()
	{
		if (_martingaleLevel == 0 || _lastTradeVolume <= 0m)
		{
			return Volume;
		}

		return _lastTradeVolume * 2m;
	}

	private decimal CalculateTotalProfit(decimal price)
	{
		var unrealized = Position != 0 ? Position * (price - PositionPrice) : 0m;
		return PnL + unrealized;
	}

	private void UpdateStepSettings()
	{
		var priceStep = Security?.PriceStep ?? 0m;
		if (priceStep <= 0m)
		{
			priceStep = 1m;
		}

		_stepSize = StepPoints * priceStep;
		_entryOffset = EntryOffsetPoints * priceStep;
	}

	private void CancelActiveOrders()
	{
		CancelOrderIfActive(ref _buyStopOrder);
		CancelOrderIfActive(ref _sellStopOrder);
		CancelOrderIfActive(ref _reversalOrder);
	}

	private void CancelOrderIfActive(ref Order order)
	{
		if (order == null)
			return;

		if (IsOrderActive(order))
		{
			CancelOrder(order);
		}

		order = null;
	}

	private static bool IsOrderActive(Order order)
	{
		if (order == null)
		{
			return false;
		}

		return order.State == OrderStates.Active ||
			order.State == OrderStates.Pending ||
			order.State == OrderStates.PartiallyFilled ||
			order.State == OrderStates.None;
	}

	private void ResetCycle()
	{
		_martingaleLevel = 0;
		_lastTradePrice = 0m;
		_lastTradeVolume = 0m;
		_lastTradeSide = null;
		_processedOrders.Clear();
		_reversalOrder = null;
	}
}

File: API/2653_Martin_Martingale/PY/martin_martingale_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Strategies import Strategy


class martin_martingale_strategy(Strategy):
    """Martingale grid that alternates long/short entries while doubling volume."""

    def __init__(self):
        super(martin_martingale_strategy, self).__init__()

        self._step_points = self.Param("StepPoints", 10) \
            .SetGreaterThanZero() \
            .SetDisplay("Step (points)", "Distance multiplier for reversals", "General")
        self._entry_offset_points = self.Param("EntryOffsetPoints", 10) \
            .SetGreaterThanZero() \
            .SetDisplay("Entry Offset (points)", "Offset for initial breakout entry", "General")
        self._profit_target = self.Param("ProfitTarget", 5.0) \
            .SetGreaterThanZero() \
            .SetDisplay("Profit Target", "Total profit to close all positions", "Risk")
        self._max_level = self.Param("MaxLevel", 5) \
            .SetGreaterThanZero() \
            .SetDisplay("Max Level", "Maximum martingale levels", "Risk")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Candles for price monitoring", "Data")

        self._step_size = 0.0
        self._entry_offset = 0.0
        self._last_trade_price = 0.0
        self._last_trade_volume = 0.0
        self._martingale_level = 0
        self._last_trade_side = 0  # 0=none, 1=buy, -1=sell
        self._is_closing = False
        self._initial_price = None

    @property
    def StepPoints(self):
        return int(self._step_points.Value)
    @property
    def EntryOffsetPoints(self):
        return int(self._entry_offset_points.Value)
    @property
    def ProfitTarget(self):
        return self._profit_target.Value
    @property
    def MaxLevel(self):
        return int(self._max_level.Value)
    @property
    def CandleType(self):
        return self._candle_type.Value

    def _update_step_settings(self):
        sec = self.Security
        step = float(sec.PriceStep) if sec is not None and sec.PriceStep is not None and float(sec.PriceStep) > 0 else 1.0
        self._step_size = self.StepPoints * step
        self._entry_offset = self.EntryOffsetPoints * step

    def _reset_cycle(self):
        self._martingale_level = 0
        self._last_trade_price = 0.0
        self._last_trade_volume = 0.0
        self._last_trade_side = 0

    def OnStarted2(self, time):
        super(martin_martingale_strategy, self).OnStarted2(time)

        self._update_step_settings()
        self._reset_cycle()
        self._is_closing = False
        self._initial_price = None

        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(self.process_candle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

    def process_candle(self, candle):
        if candle.State != CandleStates.Finished:
            return

        self._update_step_settings()

        if self._step_size <= 0 or float(self.Volume) <= 0:
            return

        price = float(candle.ClosePrice)

        # If closing, flatten and wait
        if self._is_closing:
            if self.Position == 0:
                self._is_closing = False
                self._reset_cycle()
            return

        # If flat after a cycle, reset
        if self.Position == 0 and self._martingale_level > 0:
            self._reset_cycle()

        # Check profit target
        if float(self.ProfitTarget) > 0 and float(self.PnL) >= float(self.ProfitTarget) and self.Position != 0:
            self._is_closing = True
            if self.Position > 0:
                self.SellMarket()
            elif self.Position < 0:
                self.BuyMarket()
            return

        # Max level reached
        if self._martingale_level >= self.MaxLevel and self.Position != 0:
            self._is_closing = True
            if self.Position > 0:
                self.SellMarket()
            elif self.Position < 0:
                self.BuyMarket()
            return

        # Initial entry: wait for breakout from first candle
        if self._martingale_level == 0 and self.Position == 0:
            if self._initial_price is None:
                self._initial_price = price
                return

            if self._entry_offset <= 0:
                return

            if price >= self._initial_price + self._entry_offset:
                self.BuyMarket()
                self._last_trade_price = price
                self._last_trade_volume = float(self.Volume)
                self._last_trade_side = 1
                self._martingale_level = 1
                self._initial_price = None
            elif price <= self._initial_price - self._entry_offset:
                self.SellMarket()
                self._last_trade_price = price
                self._last_trade_volume = float(self.Volume)
                self._last_trade_side = -1
                self._martingale_level = 1
                self._initial_price = None
            return

        if self._last_trade_side == 0 or self._martingale_level == 0:
            return

        threshold = self._step_size

        if self._last_trade_side == 1:
            if price <= self._last_trade_price - threshold:
                self.SellMarket()
                self._last_trade_price = price
                self._last_trade_volume *= 2.0
                self._last_trade_side = -1
                self._martingale_level += 1
        else:
            if price >= self._last_trade_price + threshold:
                self.BuyMarket()
                self._last_trade_price = price
                self._last_trade_volume *= 2.0
                self._last_trade_side = 1
                self._martingale_level += 1

    def OnReseted(self):
        super(martin_martingale_strategy, self).OnReseted()
        self._reset_cycle()
        self._is_closing = False
        self._initial_price = None
        self._step_size = 0.0
        self._entry_offset = 0.0

    def CreateClone(self):
        return martin_martingale_strategy()
